### ISSUE TITLE
ES 5.4 percolator is 3 times slower than ES 1.6 percolator

### ISSUE THREAD

**Issue by dennisgorelik** [elastic/elasticsearch#25278]
https://github.com/elastic/elasticsearch/issues/23709
ElasticSearch 5.4 percolator execution time is approximately 3 times longer than ElasticSearch 1.6 percolator execution time.
(So the empirical rule is that percolator execution time grows approximately proportional to the ElasticSearch version number: 5.4/1.6 ~= 3).

That slowness is not very surprising, considering that percolator index size increased about 7 times (from 21.1 MB in ES 1.6 to 144 MB in ES 5.4).

This is our setup:
```
227,906 percolator queries (job alerts that include text query and/or location).
30 documents to percolate in a single batch.
Processor: i7-7700 (4 cores with up to 4.20 GHz)
```

Execution time for the percolator batch
```
ES 1.6:  9.095s
ES 5.4: 26.736s
```

```
CPU utilization: 100% (on both versions of ElasticSearch)
```

In the hopes of improving performance, we tried to represent percolator queries is 3 different forms:
```
1) "bool/must"
2) "bool/filter"
3) "bool/must/filter"
```
In all 3 cases percolator index size and execution time of percolate batch changed only few percent.

I see that similar issue was already reported 3 months ago and is still open:
[Geo operations in ES 5.1.2 are slower than older ES](https://github.com/elastic/elasticsearch/issues/23709)

**Comment by martijnvg** [comment]
@dennisgorelik What kind of queries are you using in your percolator queries?

The percolator in 2.x and before cached all queries into memory. In 5.x the percolator this cache was removed (it didn't really scale well) and the percolator was improved to smarter select which percolator queries need to be evaluated instead of evaluating all queries all the time.  The selecting of candidate percolate queries is done based on query term extracting done at index time. Unfortunately we can't yet extract terms from all queries (range and wildcard queries for example), so in that case these queries will still always need to be evaluated.

**Comment by dennisgorelik** [comment]
@martijnvg - below are 10 examples of our percolator queries (out of 227906 total we have).
As you can see - we do not use wildcards, but we use boolean queries.
We use geo filters. Usually we specify exact lat/lng coordinates + distance, but we also use country&state exact match.
Do you consider such queries - a range query?
```
{
  "took" : 13,
  "timed_out" : false,
  "_shards" : {
    "total" : 8,
    "successful" : 8,
    "failed" : 0
  },
  "hits" : {
    "total" : 227906,
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "72110",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "ELECTRONICS AND ENGINEER",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 10.765833,
                      "lon" : 78.714722
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071540
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "103943",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "ALUMINIUM",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 19.707887,
                      "lon" : 83.366332
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071541
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "103984",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "MANAGEMENT AND PROJECT",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "term" : {
                    "Country" : "IN"
                  }
                },
                {
                  "term" : {
                    "State" : "WB"
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071541
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "103986",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "SIS",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "103989",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "HSE AND SAFETY",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 36.319356,
                      "lon" : -115.022874
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "104007",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "(DESKTOP AND TECH AND SUPPORT) OR (NETWORK AND ENGINEER)",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 33.956215,
                      "lon" : -83.987962
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "104027",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "OIL AND GAS AND ENERGY AND PARTS AND PROFESSIONAL",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 40.063582,
                      "lon" : -75.078636
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "104029",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "MANANGER",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 33.785795,
                      "lon" : -118.264357
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "104039",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "SPECIAL AND EDUCATION AND TEACHERS",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 39.930668,
                      "lon" : -75.320188
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      },
      {
        "_index" : "percolatejobs",
        "_type" : "jobalert",
        "_id" : "104042",
        "_score" : 1.0,
        "_source" : {
          "query" : {
            "bool" : {
              "must" : [
                {
                  "query_string" : {
                    "fields" : [
                      "JobTitle",
                      "JobDescription",
                      "CompanyName",
                      "Salary"
                    ],
                    "query" : "VITAE",
                    "analyzer" : "pjfqueryanalyzer"
                  }
                },
                {
                  "geo_distance" : {
                    "distance" : "50mi",
                    "Location" : {
                      "lat" : 14.629796,
                      "lon" : 121.137723
                    }
                  }
                }
              ]
            }
          },
          "SystemUpdateDate" : 1497595071542
        }
      }
    ]
  }
}

```

**Comment by martijnvg** [comment]
@dennisgorelik At index time the percolator can't analyze `geo_distance` queries, so percolator queries containing these queries may always need to be evaluated by the `percolate` query. 

However in your case the `geo_distance` queries are part of a `bool` with two must clauses (this wouldn't be possible with should clauses), which means at index time during query analysis it is able to ignore the unsupported query. So it will preselect these queries based on the other must clause containing the `query_string` query. This will still allow the percolator to skip over a lot of percolator queries.

Can you run the query that is described here:
https://www.elastic.co/guide/en/elasticsearch/reference/5.4/query-dsl-percolate-query.html#_how_it_works_under_the_hood

This will tell for how many queries the extraction process at index time was failed and indicates how many percolator queries always need to be evaluated when percolating.

**Comment by dennisgorelik** [comment]
@martijnvg 
It looks like only 724 alerts (out of 227907 total) were marked for mandatory percolation:
```
GET /_search
{
  "query": {
    "term" : {
      "query.extraction_result" : "failed"
    }
  }
}
```

```
{
   "took": 0,
   "timed_out": false,
   "_shards": {
      "total": 20,
      "successful": 20,
      "failed": 0
   },
   "hits": {
      "total": 724,
      "max_score": 5.895347,
      "hits": [
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "13180",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "15mi",
                                    "Location": {
                                       "lat": 26.976171,
                                       "lon": -82.090645
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014903959
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "1457",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 19.8853,
                                       "lon": 75.319656
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014903997
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "10240",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 26.846511,
                                       "lon": 80.946683
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904008
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "3802",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 26.720249,
                                       "lon": 88.424583
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904015
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "16858",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 34.540024,
                                       "lon": -112.468503
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904105
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "221",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 35.686975,
                                       "lon": -105.937799
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904260
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "23996",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 33.68906,
                                       "lon": -78.886694
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904319
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "27614",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 38.36119,
                                       "lon": -97.654811
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904380
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "32521",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "25mi",
                                    "Location": {
                                       "lat": 33.68906,
                                       "lon": -78.886694
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904468
            }
         },
         {
            "_index": "percolatejobs",
            "_type": "jobalert",
            "_id": "42273",
            "_score": 5.895347,
            "_source": {
               "query": {
                  "bool": {
                     "must": {
                        "match_all": {}
                     },
                     "filter": {
                        "bool": {
                           "must": [
                              {
                                 "geo_distance": {
                                    "distance": "50mi",
                                    "Location": {
                                       "lat": 18.050193,
                                       "lon": 78.2647
                                    }
                                 }
                              }
                           ]
                        }
                     }
                  }
               },
               "SystemUpdateDate": 1498014904646
            }
         }
      ]
   }
}
```

**Comment by martijnvg** [comment]
Thanks @dennisgorelik, then I think in your case the preselection isn't excluding a lot of percolator queries. This can be in case a document gets percolated that contains terms that most of your percolator queries have.

Maybe the cache that the percolator had before should be brought back in a different form. (it should definitively not be unbounded like it was before)

**Comment by martijnvg** [comment]
Closed in favour of #25445
